homeassistant/__main__.py

```diff
@@ -256,42 +250,61 @@ def open_browser(event):
-        sys.exit(exit_code)
     return exit_code
 
 
-def run_hass_process(hass_proc):
-    """Run a child hass process. Returns True if it should be restarted."""
-    requested_stop = threading.Event()
-    hass_proc.daemon = True
+def try_to_restart():
+    """Attempt to clean up state and start a new homeassistant instance."""
+    # Things should be mostly shut down already at this point, now just try
+    # to clean up things that may have been left behind.
+    sys.stderr.write('Home Assistant attempting to restart.\n')
+
+    # Count remaining threads, ideally there should only be one non-daemonized
+    # thread left (which is us). Nothing we really do with it, but it might be
+    # useful when debugging shutdown/restart issues.
+    nthreads = sum(thread.isAlive() and not thread.isDaemon()
+                   for thread in threading.enumerate())
+    if nthreads > 1:
+        sys.stderr.write("Found {} non-daemonic threads.\n".format(nthreads))
+
+    # Send terminate signal to all processes in our process group which
+    # should be any children that have not themselves changed the process
+    # group id. Don't bother if couldn't even call setpgid.
+    if hasattr(os, 'setpgid'):
+        sys.stderr.write("Signalling child processes to terminate...\n")
+        os.kill(0, signal.SIGTERM)
+
+        # wait for child processes to terminate
+        try:
+            while True:
+                time.sleep(1)
+                if os.waitpid(0, os.WNOHANG) == (0, 0):
+                    break
+        except OSError:
+            pass
 
-    def request_stop(*args):
-        """Request hass stop, *args is for signal handler callback."""
-        requested_stop.set()
-        hass_proc.terminate()
+    elif os.name == 'nt':
+        # Maybe one of the following will work, but how do we indicate which
+        # processes are our children if there is no process group?
+        # os.kill(0, signal.CTRL_C_EVENT)
+        # os.kill(0, signal.CTRL_BREAK_EVENT)
+        pass
 
-    try:
-        signal.signal(signal.SIGTERM, request_stop)
-    except ValueError:
-        print('Could not bind to SIGTERM. Are you running in a thread?')
+    # Try to not leave behind open filedescriptors with the emphasis on try.
+    os.closerange(3, 1024)
 
-    hass_proc.start()
-    try:
-        hass_proc.join()
-    except KeyboardInterrupt:
-        request_stop()
-        try:
-            hass_proc.join()
-        except KeyboardInterrupt:
-            return False
-
-    return (not requested_stop.isSet() and
-            hass_proc.exitcode == RESTART_EXIT_CODE,
-            hass_proc.exitcode)
+    # Now launch into a new instance of Home-Assistant. If this fails we
+    # fall through and exit with error 100 (RESTART_EXIT_CODE) in which case
+    # systemd will restart us when RestartForceExitStatus=100 is set in the
+    # systemd.service file.
+    sys.stderr.write("Restarting Home-Assistant\n")
+    exe = get_executable()
```

Comment:
I figured if multiprocessing is part of the standard library and is not using that function, it must have a reason for it.